### DISCUSSION TITLE
Storing and syncing bookmarks in a predefined place

### DISCUSSION THREAD

**Discussion by staalman** [GarinZ/link-map#19] [Ideas]
Storing bookmarks _locally_ is good, but in general (looking at TO and what it does/doesn't).

I believe users are looking for:
1. a way to store and retrieve bookmarks in a _configurable_, predefined place, like e.g.:
   a) local disk
   b) network disk or NAS
   c) cloud storage (e.g. google drive)
   based on some sort of _profile_. 
   When using cloud-based authentication - like with a google/chrome account or OpenID - cloud storage and sharing (see sharing idea) will become versatile and generic. Could even happen on a users' NAS when using OpenID for those who are privacy focused. 
That way, usage for professional vs personal vs hobby, etc. can be separated and maintained easily (which is working great in TO) with vast amounts of URLs.
  
2. when storing and retrieving the bookmarks from a networked/cloud space, it makes sense to use **profiles** for these sessions/trees to allow usage across devices:
  a) storing and retrieving centrally managed trees/sessions
  and then the long waited need for:
  b) synchronising trees across devices or profiles (see sharing idea).

  3. Currently, I've understood the tree is stored as JSON. I would like to suggest to also look at the TO solution where you can literally copy a tree and paste it in e.g. an IntelliJ window as tabbed columns text.  For example, a TO group with one TO window node, with one URL (newtab) would like look the following:
-- cut --
Group
    Window
        New Tab (chrome://newtab/)
-- cut --

That way, a delta comparison of changes between two large trees can be made available easily, which allow the manually consolidation of (versions of) trees that are both large and deep and/or from different sources. I can use IntelliJ's comparison features for that. Would be - hopefully - a temporary solution, because:

4. Ideally/eventually ( ;-) )  the tool supports tree version comparison, diff-ing and merging Out of the box, build-in. 
It is especially effective when tagging is implemented to merge trees from different sources/profiles. 
Using multiple devices with the same cloud-based link (URL)-set, fully kept in sync even when run in parallel, becomes then finally a standard feature. I'm waiting for it for around a decade already! :)

I will certainly have a peak at the source code one of these days to see if and if so how I could contribute, not certain what language knowledge is most useful?Will also point it out to others.

**Comment by GarinZ** [comment]
1. Configurable, predefined storage: this feature is in my plan and scheduled.
2. Sync across devices: this feature is in my plan and scheduled.
3. Copy subtree in tabbed format: I will add this feature in next version, it's very useful.
4. Tree diffing/merging: I have writed it in the idea section of link-map's Road Map, and will think about this feature after online sync feature released.

**Reply by GarinZ** [reply]
The technology stack of this project is:
- Language
  -  TypeScript
  - JavaScript(ES6+)
  - Less
- Framework & Lib
  - [fancytree](https://github.com/mar10/fancytree): The core lib of tree view.
  - React
